Empty state button rules

MadsBuchmann opened the pull request:
This PR closes #1222. 

## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] cookbook application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #1222 

## What is the new behavior?
When a `emptyStateComponent` has `button`-children all of them will have their attention level set to 3; except the first one with attention level 1. This also applies if buttons are dynamically added or removed from the `emptyStateComponent`. 

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information
This is a draft as I still have to add tests for adding and removing button children from the emptyStateComponent. I however need a bit of help with this; see comments.

MadsBuchmann commented on libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts:
```diff
@@ -35,4 +40,32 @@ describe('EmptyStateComponent', () => {
+    });
+
+    it('should enforce attention level rules when a button is added', () => {
+      // TODO: implement test...
```
I could use some help on how to go about this - I've been searching for other tests that do something similar, but can't seem to find any. Any help would be much appreciated 🥳

MadsBuchmann commented on libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts:
```diff
@@ -35,4 +40,32 @@ describe('EmptyStateComponent', () => {
+    });
+
+    it('should enforce attention level rules when a button is removed', () => {
+      // TODO: implement test...
```
I could use some help on how to go about this - I've been searching for other tests that do something similar, but can't seem to find any. Any help would be much appreciated 🥳

jakobe commented on apps/cookbook/src/app/examples/empty-state-example/empty-state-example.component.html:
```diff
@@ -42,3 +42,12 @@
+<kirby-empty-state
+  iconName="help"
+  title="Button attention levels"
+  subtitle="All slotted buttons will have attention level 3. Except the first button with level 1. "
```
Whitespace 😋 
```suggestion
  subtitle="All slotted buttons will have attention level 3. Except the first button with level 1."
```

MadsBuchmann replied:
Should be resolved by bd0ef64

jakobe commented on apps/cookbook/src/app/examples/empty-state-example/empty-state-example.component.html:
```diff
@@ -42,3 +42,12 @@
+<kirby-empty-state
+  iconName="help"
+  title="Button attention levels"
+  subtitle="All slotted buttons will have attention level 3. Except the first button with level 1. "
```
Consider wording:
```suggestion
  subtitle="All slotted buttons will be rendered with attention level 3, except the first button if it is configured with attention level 1. "
```

MadsBuchmann replied:
Should be resolved by bd0ef64

jakobe commented on libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts:
```diff
@@ -22,6 +24,9 @@ describe('EmptyStateComponent', () => {
       subtitle="You don't have any items. Call support to add some items to your account."
     >
       <button kirby-button>Call support</button>
+      <button kirby-button attentionLevel='2'> Mail support</button>
+      <button kirby-button attentionLevel='3'> Get directions </button>
+      <button kirby-button attentionLevel='1'> Mail support</button>
```
Whitespace 😋 
```suggestion
      <button kirby-button attentionLevel='2'>Mail support</button>
      <button kirby-button attentionLevel='3'>Get directions</button>
      <button kirby-button attentionLevel='1'>Mail support</button>
```

MadsBuchmann replied:
Should be resolved by 6703cee

jakobe commented on libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts:
```diff
@@ -35,4 +40,32 @@ describe('EmptyStateComponent', () => {
+      buttons = spectator.queryAll(ButtonComponent);
+    });
+
+    it('should change all buttons to attention level 3 except 1', () => {
```
Consider wording:
```suggestion
    it('should change all buttons to attention level 3 except the 1st', () => {
```

MadsBuchmann replied:
Should be resolved by 6703cee

jakobe commented on libs/designsystem/src/lib/components/empty-state/empty-state.component.ts:
```diff
@@ -1,13 +1,58 @@
+    this.enforceAttentionLevelRules();
+
+    this.slottedButtons.changes
+      .pipe(takeUntil(this.ngUnsubscribe$), delay(0))
```
No need to unsubscribe as this is handled by Angular when the component gets destroyed
```suggestion
```

MadsBuchmann replied:
Should be resolved by 867ba3e

jakobe commented on libs/designsystem/src/lib/components/empty-state/empty-state.component.ts:
```diff
@@ -1,13 +1,58 @@
+      .subscribe(() => this.enforceAttentionLevelRules());
+  }
+
+  ngOnDestroy() {
+    this.ngUnsubscribe$.next();
+    this.ngUnsubscribe$.complete();
+  }
+
```
No need to unsubscribe as this is handled by Angular when the component gets destroyed
```suggestion
```

MadsBuchmann replied:
Should be resolved by 867ba3e

jakobe commented on libs/designsystem/src/lib/components/empty-state/empty-state.component.ts:
```diff
@@ -1,13 +1,58 @@
+    this.ngUnsubscribe$.complete();
+  }
+
+  /** Enforces that all buttons will have their attention
+   * level set to 3 except the first encountered button
+   * with level 1
+   */
+  private enforceAttentionLevelRules() {
+    let encounteredLvl1Button = false;
+
+    this.slottedButtons.forEach((button) => {
+      if (button.isAttentionLevel1 && !encounteredLvl1Button) {
+        encounteredLvl1Button = true;
+      } else if (!button.isAttentionLevel3) {
+        button.attentionLevel = '3';
+      }
+    });
+  }
```
The [issue/tech refinement](https://github.com/kirbydesign/designsystem/issues/1222#issuecomment-779841964) might not have been described well enough, but it's only supposed to be the 1st button in the list which is allowed attentionLevel1 (i.e. the wording "first of type" hinting at using `:first-of-type` pseudo selector)

MadsBuchmann replied:
Should be resolved by 3f5f00b

jakobe commented on libs/designsystem/src/lib/components/empty-state/empty-state.component.ts:
```diff
@@ -1,13 +1,58 @@
+   * level set to 3 except the first encountered button
+   * with level 1
+   */
+  private enforceAttentionLevelRules() {
```
Nice naming 👌

MadsBuchmann replied:
Thank you 🙆‍♂️

jakobe commented on libs/designsystem/src/lib/components/empty-state/empty-state.component.ts:
```diff
@@ -1,13 +1,58 @@
+    this.ngUnsubscribe$.complete();
+  }
+
+  /** Enforces that all buttons will have their attention
```
Consider wording:
```suggestion
  /** Enforces that all slotted buttons will have their attention
```

MadsBuchmann replied:
Should be solved by bd0ef64

jakobe reviewed (changes requested):
The issue/tech refinement might not have been described well enough, but it's only supposed to be the 1st button in the list which is allowed attentionLevel1 (i.e. the wording "first of type" hinting at using :first-of-type pseudo selector)
Let's talk about tests :D

MadsBuchmann commented:
I've implemented the review suggestions. 

As I am testing how buttons behave when inserted into empty-state, i have moved the tests regarding the enforcement of attention level rules to `empty-state.component.integration.spec.ts`. 

I have reverted `empty-state.component.spec.ts` as these tests now are in `empty-state.component.integration.spec.ts`.

MadsBuchmann commented on libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts:
```diff
@@ -1,8 +1,9 @@
-import { DesignTokenHelper } from './../../helpers/design-token-helper';
 import { IconComponent } from '../icon/icon.component';
+
+import { DesignTokenHelper } from './../../helpers/design-token-helper';
```
I've attempted to revert this file - but whenever I commit the imports are re-ordered automatically.

MadsBuchmann replied:
It managed to make it work!

jakobe commented on libs/designsystem/src/lib/components/empty-state/empty-state.component.integration.spec.ts:
```diff
@@ -0,0 +1,74 @@
+              title="No items"
+              subtitle="You don't have any items. Call support to add some items to your account."
+          >
+              <button kirby-button attentionLevel='4'>Call support</button>
+              <button kirby-button attentionLevel='3'>Mail support</button>
+              <button kirby-button attentionLevel='2'>Get directions</button>
+              <button kirby-button attentionLevel='1'>Cancel</button>
```
We generally use `"` for property assignment:
```suggestion
              <button kirby-button attentionLevel="4">Call support</button>
              <button kirby-button attentionLevel="3">Mail support</button>
              <button kirby-button attentionLevel="2">Get directions</button>
              <button kirby-button attentionLevel="1">Cancel</button>
```

jakobe commented on libs/designsystem/src/lib/components/empty-state/empty-state.component.integration.spec.ts:
```diff
@@ -0,0 +1,74 @@
+              title="No items"
+              subtitle="You don't have any items. Call support to add some items to your account."
+          >
+              <button kirby-button attentionLevel='4'>Call support</button>
+              <button kirby-button attentionLevel='3'>Mail support</button>
+              <button kirby-button attentionLevel='2'>Get directions</button>
+              <button kirby-button attentionLevel='1'>Cancel</button>
```
Consider not having attentionlevel set in the default test and move this to a separate test...

MadsBuchmann replied:
I have moved them out to a separate test case `describe('EmptyStateComponent with slotted buttons configured with attention levels', () => {...` and the case you mention now uses defaults 😄

jakobe commented on libs/designsystem/src/lib/components/empty-state/empty-state.component.ts:
```diff
@@ -1,13 +1,39 @@
+  ngAfterContentInit() {
+    this.enforceAttentionLevelRules();
+
+    this.slottedButtons.changes.pipe(delay(0)).subscribe(() => this.enforceAttentionLevelRules());
```
Why the need for `delay`? Consider adding comment...

MadsBuchmann replied:
Is this sufficient 😄?

jakobe replied:
Yes much better :)
Consider stating why it's necessary as we do in RadioGroupComponent:
`// setTimeout prevents ExpressionChangedAfterItHasBeenCheckedError when updating the DOM in QueryList.changes:`
You could also consider wrapping the call to `this.enforceAttentionLevelRules` in a timeout instead, as this is the conventional way to postpone a state change to next CD...
See: https://github.com/kirbydesign/designsystem/blob/master/libs/designsystem/src/lib/components/radio/radio-group/radio-group.component.ts#L201

MadsBuchmann replied:
I've updated the comment to be: 
`/* setTimeout prevents ExpressionChangedAfterItHasBeenCheckedError when changing attention 
     levels of slotted buttons in this.enforceAttentionLevelRules */`

and removed the use of delay 😀
